GeneralJS/balanced-parens-2.ts

```diff
@@ -16,6 +16,37 @@
  *
  */
 
-function balancedParens2(input): boolean {}
+function balancedParens2(input) {
+    /*
+        ((())[]{}) = true
+        var hubble = function() { telescopes.awesome();) = false
+    */
+   
+    // breaking case : closed symbol is proceeded by a none matching open symbol
 
-module.exports = balancedParens2;
+    // define regex
+    const openRegex = /[\[\{\(]/g
+    const closedRegex = /[\]\}\)]/g
+    // keep up with pairs 
+    let parenStack : string[] = []; 
+
+    // iterate over parens
+    for(let i = 0; i < input.length; i++) {
+        // if open symbol, add to stack
+        if(input[i].match(openRegex)){
+            parenStack.push(input[i])
+        } else if(input[i].match(closedRegex)) { // if closed symbol, evaluate for break
+            if(parenStack.length == 0) return false; // not preceded by an open symbol, immediately no
+            // not proceeded by a matching symbol
+            if(input[i] == ')' && (parenStack[parenStack.length - 1] == '{' || parenStack[parenStack.length - 1] == '[')) return false
```

Comment:
This series of if statements is the only part of this code that is difficult to maintain right now, I know it seems a little counterintuitive to have both open and closed regexes and a dictionary but a dictionary would remove a lot of the pain here

https://github.com/aflinky/Algos/blob/b310156aa8ba68b35d7672115467a22a9919cea7/GeneralJS/balanced-parens-2.ts#L39-L42